// actions-toolkit/lib/exit.js
"use strict";
// tslint:disable:no-console
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The code to exit an action with a "success" state
 */
exports.SuccessCode = 0;
/**
 * The code to exit an action with a "failure" state
 */
exports.FailureCode = 1;
/**
 * The code to exit an action with a "neutral" state
 */
exports.NeutralCode = 78;
var Exit = /** @class */ (function () {
    function Exit() {
    }
    /**
     * Stop the action with a "success" status
     */
    Exit.prototype.success = function (message) {
        if (message)
            console.log(message);
        process.exit(exports.SuccessCode);
    };
    /**
     * Stop the action with a "neutral" status
     */
    Exit.prototype.neutral = function (message) {
        if (message)
            console.log(message);
        process.exit(exports.NeutralCode);
    };
    /**
     * Stop the action with a "failed" status
     */
    Exit.prototype.failure = function (message) {
        if (message)
            console.error(message);
        process.exit(exports.FailureCode);
    };
    return Exit;
}());
exports.Exit = Exit;
